lib/checks/color/link-in-text-block-evaluate.js

```diff
@@ -37,57 +37,58 @@ function linkInTextBlockEvaluate(node) {
 
   this.relatedNodes([parentBlock]);
 
-  // TODO: Check the :visited state of the link
   if (elementIsDistinct(node, parentBlock)) {
     return true;
-  } else {
-    // Check if contrast of foreground is sufficient
-    var nodeColor, parentColor;
-    nodeColor = getForegroundColor(node);
-    parentColor = getForegroundColor(parentBlock);
+  }
+
+  // Capture colors, exiting early in case of error
+  var nodeColor = getForegroundColor(node);
+  var parentColor = getForegroundColor(parentBlock);
+
+  if (!nodeColor || !parentColor) {
+    return undefined;
+  }
 
-    if (!nodeColor || !parentColor) {
-      return undefined;
-    }
+  var nodeBackgroundColor = getBackgroundColor(node);
+  var parentBackgroundColor = getBackgroundColor(parentBlock);
 
-    var contrast = getContrast(nodeColor, parentColor);
-    if (contrast === 1) {
-      return true;
-    } else if (contrast >= 3.0) {
-      incompleteData.set('fgColor', 'bgContrast');
-      this.data({
-        messageKey: incompleteData.get('fgColor')
-      });
-      incompleteData.clear();
-      // User needs to check whether there is a hover and a focus style
-      return undefined;
-    }
+  if (!nodeBackgroundColor || !parentBackgroundColor) {
+    var reason = incompleteData.get('bgColor') ?? 'bgContrast';
+    this.data({
+      messageKey: reason
+    });
+    incompleteData.clear();
+    return undefined;
+  }
 
-    // Check if contrast of background is sufficient
-    nodeColor = getBackgroundColor(node);
-    parentColor = getBackgroundColor(parentBlock);
+  // Compute ratios and return pass or fail, as appropriate
+  var textContrast = getContrast(nodeColor, parentColor);
+  var backgroundContrast = getContrast(
+    nodeBackgroundColor,
+    parentBackgroundColor
+  );
 
-    if (
-      !nodeColor ||
-      !parentColor ||
-      getContrast(nodeColor, parentColor) >= 3.0
-    ) {
-      let reason;
-      if (!nodeColor || !parentColor) {
-        reason = incompleteData.get('bgColor');
-      } else {
-        reason = 'bgContrast';
-      }
-      incompleteData.set('fgColor', reason);
-      this.data({
-        messageKey: incompleteData.get('fgColor')
-      });
-      incompleteData.clear();
-      return undefined;
-    }
+  if (textContrast >= 3.0 || backgroundContrast >= 3.0) {
+    return true;
+  }
+
+  // Message reflects the highest contrast ratio since it's the smallest fix
+  if (textContrast >= backgroundContrast) {
```

Comment:
I see where you're coming from, but I feel this may not be the right direction. For a link with 1. no underline, and 2. bad contrast, this is going to end up telling it to improve the contrast. I think what might be better is that we pull the contrast testing into a separate check, and make those "any" checks, so we can report it like so:

> Fix one of the following:
> - The link has has no styling (such as underline) to distinguish it from the surrounding text
> - The link text insufficient contrast of 2:1 with the surrounding text. (Minimum contrast 3:1, link text: #ABC123, surrounding text: #CDE234)

For the background, I think that can be combined with the text color check. People don't use backgrounds on inline links all too often, we don't have to explicitly call it out unless there is a background on the link IMO. 

Alternatively, we could do the background and foreground thing in the same check, and report whichever has the highest number like you're doing here. Changing background colors on inline links is rare, so it may not be necessary to point that option out unless someone is already using this. Might help on performance a bit too.

Re performance; When splitting this check it'll be good to do some performance testing, make sure we don't significantly slow down this rule. Color contrast can be expensive after all. If it does slow down we'll probably have to look at doing some more caching / memoization.

Reply:
Refactored to use any. I'll reach out to you offline for details on how you generally handle perf testing